src/Kurrent.Client/Core/Serialization/MessageSerializer.cs

```diff
@@ -0,0 +1,151 @@
+
+		var settings = defaultSettings.Clone();
+		operationSettings.ConfigureSettings.Invoke(settings);
+
+		return new MessageSerializer(SchemaRegistry.From(settings));
+	}
+}
+
+public class MessageSerializer(SchemaRegistry schemaRegistry) : IMessageSerializer {
+	readonly ISerializer _jsonSerializer =
+		schemaRegistry.GetSerializer(ContentType.Json);
+
+	readonly IMessageTypeNamingStrategy _messageTypeNamingStrategy =
+		schemaRegistry.MessageTypeNamingStrategy;
+
+	public EventData Serialize(Message message, MessageSerializationContext serializationContext) {
+		var (data, metadata, eventId) = message;
+
+		var eventType = _messageTypeNamingStrategy
+			.ResolveTypeName(
+				message.Data.GetType(),
+				new MessageTypeNamingResolutionContext(serializationContext.CategoryName)
+			);
+
+		var serializedData = schemaRegistry
+			.GetSerializer(serializationContext.ContentType)
+			.Serialize(data);
+
+		var serializedMetadata = metadata != null
+			? _jsonSerializer.Serialize(metadata)
+			: ReadOnlyMemory<byte>.Empty;
+
+		return new EventData(
+			eventId ?? Uuid.NewUuid(),
+			eventType,
+			serializedData,
+			serializedMetadata,
+			serializationContext.ContentType.ToMessageContentType()
+		);
+	}
+
+#if NET48
+	public bool TryDeserialize(EventRecord record, out Message? deserialized) {
+#else
+	public bool TryDeserialize(EventRecord record, [NotNullWhen(true)] out Message? deserialized) {
+#endif
+		if (!TryResolveClrType(record, out var clrType)) {
+			deserialized = null;
+			return false;
+		}
+
+		var data = schemaRegistry
+			.GetSerializer(FromMessageContentType(record.ContentType))
+			.Deserialize(record.Data, clrType!);
+
+		if (data == null) {
+			deserialized = null;
+			return false;
+		}
+
+		object? metadata = record.Metadata.Length > 0 && TryResolveClrMetadataType(record, out var clrMetadataType)
+				? _jsonSerializer.Deserialize(record.Metadata, clrMetadataType!)
+				: null;
+
+		deserialized = Message.From(data, metadata, record.EventId);
+		return true;
+	}
+
+	public static MessageSerializer From(KurrentClientSerializationSettings? settings = null) {
+		settings ??= KurrentClientSerializationSettings.Default();
+
+		return new MessageSerializer(SchemaRegistry.From(settings));
+	}
+
+	bool TryResolveClrType(EventRecord record, out Type? clrType) =>
+		schemaRegistry
+			.MessageTypeNamingStrategy
+			.TryResolveClrType(record.EventType, out clrType);
+
+	bool TryResolveClrMetadataType(EventRecord record, out Type? clrMetadataType) =>
+		schemaRegistry
+			.MessageTypeNamingStrategy
+			.TryResolveClrMetadataType(record.EventType, out clrMetadataType);
+}
+
+public class NulloMessageSerializer : IMessageSerializer {
+	public static readonly NulloMessageSerializer Instance = new NulloMessageSerializer();
+
+	public EventData Serialize(Message value, MessageSerializationContext context) {
+		throw new InvalidOperationException("Cannot serialize, automatic deserialization is disabled");
+	}
+
+#if NET48
```

Comment:
This pragma is not needed if the project is compiled with latest SDK

Reply:
I actually have the latest version, and its required indeed.
![Screenshot 2025-02-21 at 11 25 56](https://github.com/user-attachments/assets/6a352ce2-c10d-42b1-b7c1-b129e9ca0221)

Reply:
Yes, based on my findings, `NotNullWhen` is NetStandard 2.1 per: https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullwhenattribute?view=net-9.0#applies-to

and NetStandard 2.1 doesn't support .NET Framework, per: https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-1.